File: src/main/java/com/nanaios/polygonal_tech/network/PolygonalTechNetwork.java
package com.nanaios.polygonal_tech.network;

import com.nanaios.polygonal_tech.PolygonalTech;
import com.nanaios.polygonal_tech.network.packet.ClientBoundBlockEntityBufPacket;
import net.minecraftforge.network.NetworkRegistry;
import net.minecraftforge.network.simple.SimpleChannel;

public final class PolygonalTechNetwork {
    private static final String PROTOCOL_VERSION = "1";
    public static final SimpleChannel CHANNEL = NetworkRegistry.newSimpleChannel(
            PolygonalTech.rl("main"),
            () -> PROTOCOL_VERSION,
            PROTOCOL_VERSION::equals,
            PROTOCOL_VERSION::equals
    );

    private static int packetId = 0;

    private PolygonalTechNetwork() {
    }

    public static void registerMessages() {
        CHANNEL.registerMessage(
                nextPacketId(),
                ClientBoundBlockEntityBufPacket.class,
                ClientBoundBlockEntityBufPacket::encode,
                ClientBoundBlockEntityBufPacket::decode,
                ClientBoundBlockEntityBufPacket::handle
        );
    }

    private static int nextPacketId() {
        return packetId++;
    }
}

File: src/main/kotlin/com/nanaios/polygonal_tech/lib/network/PolygonalTechNetwork.kt
package com.nanaios.polygonal_tech.lib.network

import com.nanaios.polygonal_tech.lib.network.s2c.SyncValuesPacket
import com.nanaios.polygonal_tech.main.PolygonalTech
import net.minecraft.resources.ResourceLocation
import net.minecraftforge.network.NetworkRegistry
import net.minecraftforge.network.simple.SimpleChannel

object PolygonalTechNetwork {
    private const val PROTOCOL_VERSION = "1"
    private var nextPacketId = 0

    val CHANNEL: SimpleChannel = NetworkRegistry.newSimpleChannel(
        ResourceLocation.fromNamespaceAndPath(PolygonalTech.MOD_ID, "main"),
        { PROTOCOL_VERSION },
        PROTOCOL_VERSION::equals,
        PROTOCOL_VERSION::equals
    )

    fun register() {
        CHANNEL.messageBuilder(SyncValuesPacket::class.java,nextPacketId++)
            .encoder(SyncValuesPacket::encode)
            .decoder(SyncValuesPacket::decode)
            .consumerMainThread(SyncValuesPacket::handle)
            .add()
    }
}